course/api/views.py

```diff
@@ -195,20 +198,60 @@ class CourseExercisesViewSet(NestedViewSetMixin,
     lookup_url_kwarg = 'exercisemodule_id'
     lookup_value_regex = REGEX_INT
     parent_lookup_map = {'course_id': 'course_instance.id'}
-    serializer_class = CourseModuleSerializer
 
-    def get_queryset(self) -> QuerySet[CourseModule]:
-        return (
-            CourseModule.objects
-            .get_visible(self.request.user)
-            .prefetch_related(
-                Prefetch('learning_objects', BaseExercise.objects.all(), 'exercises')
-            )
-            .all()
-        )
+    # this converts a single module into a response dictionary
+    def __module_to_dict(self, module, **kwargs):
+        # if specific module then detail otherwise all modules (list)
+        api_url_string = "course-exercises-detail" if 'exercisemodule_id' in kwargs else "course-exercises-list"
+        api_url = build_aplus_url(api_reverse(api_url_string, kwargs=kwargs))
+        module_html_url = build_aplus_url(module['link'])
+        module_dictionary = {
+            'id': module['id'],
+            'url': api_url,
+            'html_url': module_html_url,
+            'display_name': module['name'],
+            'is_open': CourseModule.is_open_helper(module['reading_opening_time'], module['opening_time'],
+                                                   module['closing_time']
+                                                   ),
+            'exercises': [],
+        }
+        # add this module to cached exercises
+        for exercise in filter(lambda ex: ex['type'] == 'exercise', module['children']):
```

Comment:
The original implementation excluded chapters from the output; only the (submittable) exercises were listed. This new implementation includes the chapters. I think `CachedContent` includes a field `submittable: True/False` for each `type: "exercise"` object. Chapters have `submittable False` and the exercises `True`.

Note that the exercises may be nested under the chapters as their children, but the exercises may also be listed directly under the module's children. A+ in theory supports unlimited nesting for the learning objects, thus you should traverse all levels of the "children" until the leaf objects have no children. You can see that the class `CachedContent` uses recursion for that.